net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+ /*
+ * net/sched/sch_gkprioq.c  Gatekeeper Priority Queue.
+ *
+ *		This program is free software; you can redistribute it and/or
+ *		modify it under the terms of the GNU General Public License
+ *		as published by the Free Software Foundation; either version
+ *		2 of the License, or (at your option) any later version.
+ *
+ * Authors:	Nishanth Devarajan, <ndev_2021@gmail.com>
+ *	        original idea by Michel Machado, Cody Doucette and Meng Xiang
+ *
+ */
+
+#include <linux/string.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/errno.h>
+#include <linux/skbuff.h>
+#include <net/pkt_sched.h>
+#include <net/sch_generic.h>
+#include <net/dsfield.h>
+#include <net/inet_ecn.h>
+#include <asm/byteorder.h>
+
+/* Priority markers */
+#define MAX_PRIORITY 63
+#define HALF_MAX_PRIORITY 31
+
+
+
+/*	  Gatekeeper Priority Queue
+ *	=================================
+ *
+ *
+ * Gatekeeper is an open source DoS defense software which sends traffic based   * on an assigned priority for the client's requeust. Assigning priorities to    * packets enables us to give low priority to clients who quickly send many      * requests, so  that we can service legitimate clients (who typically send      * requests more slowly). The DoS scenario is when assigning low priorities      * helps block out malicious senders, who could otherwise overwhelm Gatekeeper   * with a very high number of requests, denying service to the requests of 
+ * legitimate users.
+ *
+ *
+ * To assign priority to packet, Gatekeeper uses the six bits of the DSCP field, * as it is a sufficient and convenient way to limit the maximum number of       * priorities we'd like to deal with. Additionally, using the DSCP field also    * allows us to more easily use tools like iptables and tc, which already        * understand fields in the IP header. 
+ *
+ *
+ * Although the DSCP field is usually used to specify DiffServe traffic 
+ * classifcation options, it is NOT used for that intent here. 
+ *
+ *
+ *
+ *   The Gatekeper standalone codebase is found here: 
+ *	          
+ *          https://github.com/AltraMayor/gatekeeper
+ *
+ *
+ */
+
+
+struct gkprioq_sched_data {
+/* Parameters */
+	u32 max_limit;
+
+/*Variables*/
+	struct Qdisc *prio_array[MAX_PRIORITY + 1];
+	u16 last_dropped_prio;
+	u16 highest_prio;
+	u32 init_limit;
+};
+
+static inline u32 calc_alloc_len(const struct gkprioq_sched_data *q)
+{
+	u32 total_alloc_len = 0;
+	int prio;
+	
+	for (prio = 0; prio < MAX_PRIORITY + 1; prio++) 
+		total_alloc_len += (q->prio_array[prio])->limit;
+
+	return total_alloc_len;
+}
+
+static inline u16 calc_new_high_prio(const struct gkprioq_sched_data *q)
+{
+	int prio;
+
+	if (q->highest_prio > (HALF_MAX_PRIORITY + 1)) {
```

Comment:
Why did you break the scan in two parts? Why not scanning from MAX_PRIORITY to 0 in a single loop?